### PR TITLE
Improve the computational efficiency of the Bspline interpolation

### PR DESCRIPTION
**Pull request by NiklasKotarsky**
## Main changes of this PR

Change the matrix A which contains the time step information in the Bspline algorithm to sparse.

## Motivation and additional information

The current implementation of the Bspline algorithm uses a dense matrix as A containing the time steps, which results in a significant slow down a large number of time steps. This change utilizes the sparse nature of the matrix A and thus speeds up the considerably for a large number of time steps. 

## Author's checklist

* [x] I used the [`pre-commit` hook](https://precice.org/dev-docs-dev-tooling.html#setting-up-pre-commit) to prevent dirty commits and used `pre-commit run --all` to format old commits.
* [x] I added a changelog file with `make changelog` if there are user-observable changes since the last release.
* [x] I added a test to cover the proposed changes in our test suite.
* [x] For breaking changes: I documented the changes in the appropriate [porting guide](https://precice.org/couple-your-code-porting-overview.html).
* [x] I sticked to C++17 features.
* [x] I sticked to CMake version 3.22.1.
* [x] I squashed / am about to squash all commits that should be seen as one.

## Reviewers' checklist

* [ ] Does the changelog entry make sense? Is it formatted correctly?
* [ ] Do you understand the code changes?

### CODE REVIEW

**Comment by uekerman** [comment]
Looks interesting. Could you report on example performance data? How does it behave for only a few time steps per window?

**Review by BenjaminRodenberg — APPROVED** [review]
Changes look reasonable to me. @fsimonis do you have any readability advice? Looks ok to me, but maybe you have some nice ideas.

I also compared the old (`develop`) and new (#2021) implementation with the tutorial case I'm contributing via https://github.com/precice/tutorials/pull/500. If I use a large time window + strict tolerances for the time stepper, I get already observable overhead with 370 substeps for the `develop` version, with 1720 substeps the overhead gets pretty much unbearable for such a simple experiment. Comparing to the #2021 version I see a clear improvement. With 370 substeps there is no real overhead and with 1720 substeps the situation is much better.

I did not do a detailed profiling, but the positive effect of this PR is already quite visible for this simple case, if there are sufficiently many points.

**Comment by NiklasKotarsky** [comment]
> Looks interesting. Could you report on example performance data? How does it behave for only a few time steps per window?

I have timed the old (develop) and new (https://github.com/precice/precice/pull/2021) implementation using the test ReadWriteScalarDataSubcyclingThird in test/serial/time/implicit/serial-coupling, which does 4 timesteps with a third order bspline interpolation. I changed the number of time windows to 100 to get a more representative picture and used the package chrono to time the while (precice.isCouplingOngoing()) loop. 

For the current version of preCICe develop I got a time of 26.6 seconds and my changes with a sparse matrix took 26.8 seconds. 

Changing the test to do 400 timesteps and 10 time windows instead the current version of preCICe develop gets a time of 3.6 seconds and my changes with a sparse matrix took 2.9 seconds.

File: src/math/Bspline.cpp [hunk 1]
```diff
@@ -28,23 +29,29 @@ Bspline::Bspline(Eigen::VectorXd ts, const Eigen::MatrixXd &xs, int splineDegree
...
   for (Eigen::DenseIndex i = 1; i < n - 1; ++i) {
-    //Attempt at hack... the spline dimension is not used here explicitly
-    const Eigen::DenseIndex span = Eigen::Spline<double, 1>::Span(ts[i], splineDegree, _knots);
+    const Eigen::DenseIndex span      = Eigen::Spline<double, 1>::Span(ts[i], splineDegree, _knots);
+    auto                    basisFunc = Eigen::Spline<double, 1>::BasisFunctions(ts[i], splineDegree, _knots);
 
-    // The segment call should somehow be told the spline order at compile time.
-    A.row(i).segment(span - splineDegree, splineDegree + 1) = Eigen::Spline<double, 1>::BasisFunctions(ts[i], splineDegree, _knots);
+    for (Eigen::DenseIndex j = 0; j < splineDegree + 1; ++j) {
+      matrixEntries.emplace_back(Eigen::Triplet<double>(i, span - splineDegree + j, basisFunc(j)));
```
Comment (fsimonis) [inline src/math/Bspline.cpp:41]: The constructor isn't _normally_ needed to emplace. This also applies for the elements below.

```cpp
      matrixEntries.emplace_back(i, span - splineDegree + j, basisFunc(j));
```

File: src/math/Bspline.cpp [hunk 2]
```diff
@@ -28,23 +29,29 @@ Bspline::Bspline(Eigen::VectorXd ts, const Eigen::MatrixXd &xs, int splineDegree
   //The code for computing the knots and the control points is copied from Eigens bspline interpolation with minor modifications, https://gitlab.com/libeigen/eigen/-/blob/master/unsupported/Eigen/src/Splines/SplineFitting.h
 
   Eigen::KnotAveraging(ts, splineDegree, _knots);
-  Eigen::DenseIndex n = xs.cols();
-  Eigen::MatrixXd   A = Eigen::MatrixXd::Zero(n, n);
+  Eigen::DenseIndex           n = xs.cols();
+  Eigen::SparseMatrix<double> A(n, n);
```
Comment (fsimonis) [inline src/math/Bspline.cpp:33]: Could you move this underneath the loop, to the place where you use A for the first time?

File: src/math/Bspline.cpp [hunk 3]
```diff
@@ -28,23 +29,29 @@ Bspline::Bspline(Eigen::VectorXd ts, const Eigen::MatrixXd &xs, int splineDegree
...
 
+  std::vector<Eigen::Triplet<double>> matrixEntries;
```
Comment (fsimonis) [inline src/math/Bspline.cpp:35]: Given that this is a big vector, and we know the amount of elements that will be added, I am positive that reserving space will make a significant performance difference as it prevent reallocations.

```cpp
matrixEntries.reserve( /* num elements from loop n*splineDegree? */);
```

File: src/math/Bspline.cpp [hunk 4]
```diff
@@ -28,23 +29,29 @@ Bspline::Bspline(Eigen::VectorXd ts, const Eigen::MatrixXd &xs, int splineDegree
...
+    }
   }
-  A(0, 0)         = 1.0;
-  A(n - 1, n - 1) = 1.0;
 
-  auto qr = A.householderQr();
+  matrixEntries.emplace_back(Eigen::Triplet<double>(0, 0, 1.0));
+  matrixEntries.emplace_back(Eigen::Triplet<double>(n - 1, n - 1, 1.0));
+  A.setFromTriplets(matrixEntries.begin(), matrixEntries.end());
+  A.makeCompressed();
```
Comment (fsimonis) [inline src/math/Bspline.cpp:48]: It would be interesting to see the cost-reward of this based on the amount of substeps and samplings one does.

**Comment by fsimonis** [comment]
It would be interesting to use Events in the Bspline implementation and combine it with #2013.

This way, you could use the existing performance tools to analyze the cost.

File: src/math/Bspline.cpp [hunk 5]
```diff
@@ -28,23 +29,29 @@ Bspline::Bspline(Eigen::VectorXd ts, const Eigen::MatrixXd &xs, int splineDegree
...
 
-  Eigen::MatrixXd controls = Eigen::MatrixXd::Zero(n, _ndofs);
+  Eigen::SparseQR<Eigen::SparseMatrix<double>, Eigen::COLAMDOrdering<int>> qr;
+  qr.analyzePattern(A);
+  qr.factorize(A);
 
+  Eigen::MatrixXd controls = Eigen::MatrixXd::Zero(n, _ndofs);
   for (int i = 0; i < _ndofs; i++) {
     controls.col(i) = qr.solve(xs.row(i).transpose());
```
Comment (fsimonis) [inline src/math/Bspline.cpp:56]: Would it be possible to solve all `_ndofs` systems in one go by passing a matrix of systems to [solve](https://eigen.tuxfamily.org/dox/classEigen_1_1SparseQR.html#aa77f5cb833bbea615adb532139761971)?

Something like this:

```cpp
_ctrls = qr.solve(xs);
```

**Comment by BenjaminRodenberg** [comment]
One thing I just realized: Please don't forget to add a changelog entry before merging. But I can also do another review loop when you are done with the remarks of @fsimonis.

**Comment by fsimonis** [comment]
@NiklasKotarsky this looks pretty neat now!

Could you please compare what the final difference is between develop and this current version? :pray: :eyes:

File: docs/changelog/2021.md [hunk 6]
```diff
@@ -0,0 +1 @@
+- Speed up the creation of the Bspline interpolation, by switching the matrix A inside the Bspline class to a sparse matrix
```
Comment (fsimonis) [inline docs/changelog/2021.md:1]: Could you please reword this to start with `- Improved `?
This results in a more digestible changelog.

File: src/math/Bspline.cpp [hunk 7]
```diff
@@ -28,27 +29,31 @@ Bspline::Bspline(Eigen::VectorXd ts, const Eigen::MatrixXd &xs, int splineDegree
...
   for (Eigen::DenseIndex i = 1; i < n - 1; ++i) {
-    //Attempt at hack... the spline dimension is not used here explicitly
-    const Eigen::DenseIndex span = Eigen::Spline<double, 1>::Span(ts[i], splineDegree, _knots);
+    const Eigen::DenseIndex span      = Eigen::Spline<double, 1>::Span(ts[i], splineDegree, _knots);
+    auto                    basisFunc = Eigen::Spline<double, 1>::BasisFunctions(ts[i], splineDegree, _knots);
 
-    // The segment call should somehow be told the spline order at compile time.
-    A.row(i).segment(span - splineDegree, splineDegree + 1) = Eigen::Spline<double, 1>::BasisFunctions(ts[i], splineDegree, _knots);
+    for (Eigen::DenseIndex j = 0; j < splineDegree + 1; ++j) {
+      matrixEntries.insert(matrixEntries.end(), Eigen::Triplet<double>(i, span - splineDegree + j, basisFunc(j)));
```
Comment (fsimonis) [inline src/math/Bspline.cpp:41]: My suggestion was the following:

```suggestion
      matrixEntries.emplace_back(i, span - splineDegree + j, basisFunc(j));
```

To clarify:

`emplace_back` constructs an element in-place using the given arguments. So, for a `vector` with element type `Eigen:Triplet<double>`, calling `emplace_back(x,y,z)` is like calling `push_back(Eigen::Triplet<double>(x,y,z))`.

In this case, this is mostly a readability improvement.

**Comment by NiklasKotarsky** [comment]
> @NiklasKotarsky this looks pretty neat now!
> 
> Could you please compare what the final difference is between develop and this current version? 🙏 👀

Thanks for the review, I re-timed the new implementation (https://github.com/precice/precice/pull/2021) and for the 4 timesteps and 100 time windows I now get a runtime of 26.6 seconds, yielding similar performance to preCICe develop. For 1 time window and 2000 time steps it took 0.4 seconds compared to 3.3 seconds of preCICe develop.

File: src/math/Bspline.cpp (hunk 7 above)
Comment (NiklasKotarsky) [reply src/math/Bspline.cpp]: Thanks for the nice explanation. I originally had 'push_back((x,y,z))' without the 'Eigen::Triplet<double>', which triggered an error in the linter so I switched to 'emplace_back((x,y,z))', since this was suggested by the linter.

**Review by fsimonis — APPROVED** [review]
Let me know when you are done an I'll merge the PR. :smiley:
